### PR TITLE
NNlib activation function rules

### PR DESCRIPTION
**Pull request by AstitvaAggarwal**
Current NNlib rrules have incomplete handling for `logsumexp`, `softmax` and `logsoftmax`.

The following PR adds :

- Updated Native Mooncake rrules (current `@from_rrule` fails for Adjoint/Transpose because `increment_and_get_rdata!` overload is not present, is unsafe to implement.) 
- tests for rrules and helper methods
- Unblocks Lux.jl Integration. see https://github.com/LuxDL/Lux.jl/actions/runs/22794547300/job/66139118086?pr=1664, https://github.com/LuxDL/Lux.jl/actions/runs/22794547300/job/66139118021?pr=1664.
- future proofs current rules for common activation functions

### CODE REVIEW

**Comment by AstitvaAggarwal** [comment]
Fyi it's best to include this PR in the latest release for mooncake.

**Comment by AstitvaAggarwal** [comment]
@yebai  let's get this PR merged before https://github.com/chalk-lab/Mooncake.jl/pull/1044, it's easier to handle the merge conflicts for me in that PR

**Comment by yebai** [comment]
A few comments from Copilot. 

1. High – `logsumexp` rule can produce incorrect primal/gradients for `Inf` inputs
    - In `ext/MooncakeNNlibExt.jl:139` and `ext/MooncakeNNlibExt.jl:160`, the new rule computes `tmp = exp.(xp .- max_)` and `y = max_ + log(s)` directly.
    - For cases like `xp = [Inf, Inf]`, `xp .- max_` becomes `NaN`, so `tmp`, `s`, and `y` become `NaN`; mathematically `logsumexp([Inf, Inf])` should be `Inf`.
  
2. High – `kw logsumexp` pullback shape/seed contract looks wrong for scalar-output kw cases
    - In `ext/MooncakeNNlibExt.jl:151`, `kw logsumexp` uses a single pullback signature `logsumexp_kw_pb!!(::NoRData)` at `ext/MooncakeNNlibExt.jl:164`.
    - If kw usage yields scalar output (for example explicit full reduction), reverse mode typically needs scalar seed `dy`, not `NoRData`.
    - Existing `LogExpFunctions` extension splits colon/non-colon kw handling specifically for this reason; this `NNlib` rule does not.

3. Medium – tests do not cover the two highest-risk edge cases above
    - `NNlib` extension tests added in `test/ext/nnlib/nnlib.jl` cover `Adjoint/Transpose` and several `dims`, but not `Inf/NaN` stability or scalar-output kw reduction for `logsumexp`.
    - That leaves regressions from items 1–2 undetected.

4. High – the new `NNlib` primitives claim all `AbstractArrays`, but the reverse accumulator only supports plain arrays plus `Adjoint/Transpose`
    - The primitive declarations are broad at `ext/MooncakeNNlibExt.jl:58`, `ext/MooncakeNNlibExt.jl:94`, and `ext/
      MooncakeNNlibExt.jl:130`.
    - But `_accum_fdata!` only has methods for `AbstractArray` and `FData` with `Adjoint/Transpose` at `ext/
      MooncakeNNlibExt.jl:44`.
    - That means wrappers like `view`, `reshape`, `reinterpret`, `Diagonal`, or triangular wrappers are likely to hit `MethodError` in reverse mode once these new primitives are selected, even though `Mooncake` already has wrapper-handling machinery elsewhere such as `src/rules/blas.jl:41`.
    - The current tests only add `Adjoint/Transpose`; they do not cover those other wrappers.
    - Perhaps use `arrayify` on input arguments inside the rule?

File: ext/MooncakeNNlibExt.jl [hunk 1]
```diff
@@ -19,9 +39,135 @@ const SupportedArray{P,N} = Union{Array{P,N},AbstractGPUArray{P,N}}
...
+_accum_fdata!(xf::AbstractArray, ::AbstractArray{T}, g::T) where {T<:IEEEFloat} = xf .+= g
+function _accum_fdata!(xf::FData, ::Adjoint{T}, g::AbstractArray{T}) where {T<:IEEEFloat}
+    return xf.data.parent .+= g'
+end
+function _accum_fdata!(xf::FData, ::Transpose{T}, g::AbstractArray{T}) where {T<:IEEEFloat}
+    return xf.data.parent .+= transpose(g)
+end
+
+# logsoftmax rrules
+@is_primitive MinimalCtx Tuple{typeof(logsoftmax),AbstractArray{<:IEEEFloat}}
```
Comment (sunxd3) [inline ext/MooncakeNNlibExt.jl]: the old code is defined on `SupportedArray`, is it safe to use `AbstractArray`?

File: ext/MooncakeNNlibExt.jl [hunk 2]
```diff
@@ -19,9 +39,135 @@ const SupportedArray{P,N} = Union{Array{P,N},AbstractGPUArray{P,N}}
...
+    kw::CoDual{<:NamedTuple{(:dims,)}},
+    ::CoDual{typeof(logsumexp)},
+    x::CoDual{<:AbstractArray{T}},
+) where {T<:IEEEFloat}
+    dims = primal(kw).dims
+    xp = primal(x)
+    max_ = maximum(xp; dims, init=typemin(T))
+    @fastmath tmp = exp.(xp .- max_)
+    s = sum(tmp; dims)
+    @fastmath y = max_ .+ log.(s)
```
Comment (sunxd3) [inline ext/MooncakeNNlibExt.jl:175]: could `s` be a scalar here?

File: test/ext/nnlib/nnlib.jl [hunk 3]
```diff
@@ -118,6 +164,29 @@ dropout_tester_3(Trng, x, p) = dropout(Trng(1), x, p; dims=(1, 2))
             _rand(rng, 3, 3, 2),
         ),
 
+        # logsumexp with Adjoint, Transpose
+        (false, :stability, true, logsumexp, _rand(rng, 2, 3)'),
+        (false, :none, true, Core.kwcall, (dims=1,), logsumexp, _rand(rng, 3, 3)'),
```
Comment (sunxd3) [inline test/ext/nnlib/nnlib.jl]: why `none`?

Comment (AstitvaAggarwal) [reply test/ext/nnlib/nnlib.jl]: The kwarg rule is type unstable for `Adjoints`, `Transpose` types when `maximum(...; dims)` is called on them. Solved in the current PR state, by `collect`ing the `Adjoint`/`Transpose` so that `maximum` does not work with a permutated view of the `AbstractArray` on `CPU` (this causes type instability)

File: ext/MooncakeNNlibExt.jl (hunk 1 above)
Comment (AstitvaAggarwal) [reply ext/MooncakeNNlibExt.jl]: The point of writing Mooncake rules is to support Adjoint, Transpose and other such unique abstractarray objects. SupportedArray is not enough for many applications, ill make the current code more robust. 

Solved in latest PR state, (@yebai this also addresses your review's point 4)

File: ext/MooncakeNNlibExt.jl (hunk 2 above)
Comment (AstitvaAggarwal) [reply ext/MooncakeNNlibExt.jl:175]: never, because `sum` is called with `dims` in the kwarg rule only.  It will be scalar in the non kwarg rule only (as no dims passed). see the docstrings for `logsumexp`, it behaves like `sum` - the primal shape depends on kwarg `dim`.   @yebai this is also the response to your review's point 2

**Review by sunxd3 — APPROVED** [review]

File: ext/MooncakeNNlibExt.jl [hunk 4]
```diff
@@ -19,9 +39,156 @@ const SupportedArray{P,N} = Union{Array{P,N},AbstractGPUArray{P,N}}
     Tuple{typeof(dropout),AbstractRNG,SupportedArray{P},P} where {P<:IEEEFloat},
     true,
 )
-@from_rrule(MinimalCtx, Tuple{typeof(softmax),SupportedArray{<:IEEEFloat}}, true)
-@from_rrule(MinimalCtx, Tuple{typeof(logsoftmax),SupportedArray{<:IEEEFloat}}, true)
-@from_rrule(MinimalCtx, Tuple{typeof(logsumexp),SupportedArray{<:IEEEFloat}}, true)
+
+# Helper accumulator functions
```
Comment (yebai) [inline ext/MooncakeNNlibExt.jl]:
```suggestion
# Accumulator helper functions
```

File: ext/MooncakeNNlibExt.jl [hunk 5]
```diff
@@ -19,9 +39,156 @@ const SupportedArray{P,N} = Union{Array{P,N},AbstractGPUArray{P,N}}
...
+    return error(
+        """
+    `_accum_fdata!` is not implemented for array type $(typeof(x)).
+    If you are trying to differentiate through `softmax`, `logsoftmax`, or `logsumexp`
+    with a custom array wrapper, please open an issue or add a method:
+
+        _accum_fdata!(xf::YourTangentType, ::YourArrayType{T}, g::YourGradType) where {T<:IEEEFloat}
+    """,
+    )
+end
```
Comment (yebai) [inline ext/MooncakeNNlibExt.jl]: Can you use `arrayify` in relevant rules to avoid these helpers? See https://chalk-lab.github.io/Mooncake.jl/stable/utilities/defining_rules/#Canonicalising-Tangent-Types

```suggestion
```

File: ext/MooncakeNNlibExt.jl [hunk 6]
```diff
@@ -19,9 +39,156 @@ const SupportedArray{P,N} = Union{Array{P,N},AbstractGPUArray{P,N}}
...
+
+# logsoftmax rrules
+@is_primitive MinimalCtx Tuple{typeof(logsoftmax),AbstractArray{T}} where {T<:IEEEFloat}
```
Comment (yebai) [inline ext/MooncakeNNlibExt.jl]: It's better to limit these rules to `SupportedArray`. The same suggestions apply to all uses of `AbstractArray`. 

```suggestion
@is_primitive MinimalCtx Tuple{typeof(logsoftmax),SupportedArray{T}} where {T<:IEEEFloat}
```

File: ext/MooncakeNNlibExt.jl [hunk 7]
```diff
@@ -19,9 +39,156 @@ const SupportedArray{P,N} = Union{Array{P,N},AbstractGPUArray{P,N}}
...
+    @fastmath y = max_ + log(s)
+    res = zero_fcodual(y)
+    function logsumexp_pb!!(dy::T)
+        _accum_fdata!(tangent(x), xp, dy .* tmp ./ s)
+        return NoRData(), NoRData()
+    end
+    return res, logsumexp_pb!!
+end
+
+function Mooncake.rrule!!(
```
Comment (yebai) [inline ext/MooncakeNNlibExt.jl:163]: Let's split this rule into two rules, one for CPU and one for GPU, if the logic differs between CPU and GPU. Clarity is more important than saving a few lines of code. 

The same suggestion applies to other rules with similar patterns.

File: ext/MooncakeNNlibExt.jl (hunk 5 above)
Comment (AstitvaAggarwal) [reply ext/MooncakeNNlibExt.jl]: Unfortunately `arrayify` has no method for `Adjoint/Transpose` or other exotic `AbstractArrays`, and it falls through to the error fallback in `src/rules/blas.jl:70` which throws an error.
The `_accum_fdata!` helpers are necessary here because for example : the tangent accumulation for `Adjoint/Transpose` requires writing back into `xf.data.parent` with the appropriate adjoint/transpose operation, which `arrayify` doesn't handle.

In fact the fallback for `_accum_fdata!` is also future proof when encountering unhandled `AbstractArray` objects.

File: ext/MooncakeNNlibExt.jl (hunk 6 above)
Comment (AstitvaAggarwal) [reply ext/MooncakeNNlibExt.jl]: without `AbstractArray` we will not be able to take advantage of the future proof  `_accum_fdata!` fallback.
`SupportedArray` is inadequate to handle anything beyond `Arrays, GPUArrays` etc

**Comment by AstitvaAggarwal** [comment]
@yebai I have now handled and added tests for `Inf` Instability as mentioned in points 1, 3 [here](https://github.com/chalk-lab/Mooncake.jl/pull/1043#issuecomment-4025326659)

File: ext/MooncakeNNlibExt.jl (hunk 5 above)
Comment (yebai) [reply ext/MooncakeNNlibExt.jl]: Let's add `Adjoint/Transpose` support for `arrayify`, then add `Adjoint/Transpose`  to `SupportedArrays` types.

File: ext/MooncakeNNlibExt.jl (hunk 7 above)
Comment (AstitvaAggarwal) [reply ext/MooncakeNNlibExt.jl:163]: The GPU instability only comes in `logsumexp` due to `maximum` usage - no other rule is affected.

Im not sure i understand the CPU, GPU splitting for this rule, the current methods/operators dispatch correctly according to GPU, CPU usage? The `if` condition was only needed to avoid the `PermutedDimsArray` to be operated on in a unstable manner (`collect` fixes this to an `Array`)

**Comment by AstitvaAggarwal** [comment]
@yebai @sunxd3 should we get this (and the other PR) merged and add the adjoint, transpose arrayify functionality + integration seperately? 

I say this because the main review points seem to be addressed already, the additional changes are not trivial and extend beyond this PRs scope (3 activation functions)

**Comment by yebai** [comment]
Let's merge #1048 and #1049 into this branch first -- both are ready to go.
